### PR TITLE
Implemented local root for nativefs modules.

### PR DESCRIPTION
**Pull request by roliveri**
Enables mounting of any directory (not just /) on a mount point.

### CODE REVIEW

**Comment by roliveri** [comment]
@movitto @jerryk55 please review.

File: lib/virtfs/nativefs/thick/dir_class_methods.rb [hunk 1]
```diff
@@ -5,30 +5,30 @@ module VirtFS::NativeFS # rubocop:disable Style/ClassAndModuleChildren
...
+        VfsRealDir.mkdir(apply_root(p), permissions)
       end
 
-      def dir_new(_fs_rel_path, hash_args, open_path, cwd)
+      def dir_new(fs_rel_path, hash_args, _open_path, cwd)
         owd = VfsRealDir.getwd
         begin
           VfsRealDir.chdir(cwd)
-          return VfsRealDir.new(open_path, hash_args)
+          return VfsRealDir.new(apply_root(fs_rel_path), hash_args)
```
Comment (movitto) [inline lib/virtfs/nativefs/thick/dir_class_methods.rb:31]: @roliveri here you apply root to fs_rel_path when originally this was using open_path. Is this intended?

File: lib/virtfs/nativefs/thick/file_class_methods.rb [hunk 2]
```diff
@@ -3,172 +3,172 @@
...
+        VfsRealFile.writable_real?(apply_root(p))
       end
 
-      def file_new(_f, parsed_args, open_path, cwd)
+      def file_new(f, parsed_args, _open_path, cwd)
         owd = VfsRealDir.getwd
         begin
           VfsRealDir.chdir(cwd)
-          return VfsRealFile.new(open_path, *parsed_args.args)
+          return VfsRealFile.new(apply_root(f), *parsed_args.args)
```
Comment (movitto) [inline lib/virtfs/nativefs/thick/file_class_methods.rb:171]: @roliveri same comment as above

File: lib/virtfs/nativefs/thin/dir_class_methods.rb [hunk 3]
```diff
@@ -5,27 +5,27 @@ module VirtFS::NativeFS # rubocop:disable Style/ClassAndModuleChildren
...
       end
 
       def dir_mkdir(p, permissions)
-        VfsRealDir.mkdir(p, permissions)
+        VfsRealDir.mkdir(apply_root(p), permissions)
       end
 
       def dir_new(fs_rel_path, hash_args, _open_path, _cwd)
-        Dir.new(self, lookup_dir(fs_rel_path, hash_args), hash_args)
+        Dir.new(self, lookup_dir(apply_root(fs_rel_path), hash_args), hash_args)
```
Comment (movitto) [inline lib/virtfs/nativefs/thin/dir_class_methods.rb:28]: @roliveri here the original is fs_rel_path, is this what it's supposed to be?

File: lib/virtfs/nativefs/thin/file_class_methods.rb [hunk 4]
```diff
@@ -3,169 +3,169 @@
...
       end
 
       def file_writable_real?(p)
-        VfsRealFile.writable_real?(p)
+        VfsRealFile.writable_real?(apply_root(p))
       end
 
       def file_new(f, parsed_args, _open_path, _cwd)
-        File.new(self, lookup_file(f, parsed_args), parsed_args)
+        File.new(self, lookup_file(apply_root(f), parsed_args), parsed_args)
```
Comment (movitto) [inline lib/virtfs/nativefs/thin/file_class_methods.rb:168]: @roliveri ditto

**Review by movitto — COMMENTED** [review]
@roliveri Looked over the code, and save the comments it looks good. I didn't test it or the specs though, am in the middle of writing some docs for virtfs and integrating camcorder into the fat32 module. But when I'm at a good stopping point, I'll try this out

File: lib/virtfs/nativefs/thick/dir_class_methods.rb (hunk 1 above)
Comment (roliveri) [reply lib/virtfs/nativefs/thick/dir_class_methods.rb:31]: Yep, it's intentional. Going forward I'll have to re-evaluate the need for `open_path`.

**Pull request merged**